File: Golang/SingleChain.go
package main

/*
 *	借鉴: https://www.imooc.com/article/20582#0-tsina-1-29356-397232819ff9a47a7b7e80a40613cfe1
 *
 */
import (
	"fmt"
	"strings"
	"math"
)

type Element int
type Node struct{
	Data Element
	Next *Node   // 这里只能是指针。不是指针的话报错，Golang不支持这种嵌套
}

type SingleChain struct{
	Head *Node   // 这里如果不是引用的话，不好处理，因为Golang不好判断结构体是否为空
	Size int
}

func (s *SingleChain) Add(data *Node){
	if s.Size == 0{
		s.Head = data
		s.Size++
		return
	}

	cursor := s.Head
	for cursor.Next != nil{
		cursor = cursor.Next
	}
	cursor.Next = data
	s.Size++
}

func (s *SingleChain) IsEmpty() bool{
	return s.Head == nil
}

func (s *SingleChain) String() string{

	if s.IsEmpty(){
		return "no data"
	}else{
		result := make([]string, 0)
		cursor := s.Head
		for cursor != nil{
			result = append(result, fmt.Sprintf("%v",cursor.Data))
			cursor = cursor.Next
		}
		return strings.Join(result, "->")
	}

}

func (s *SingleChain) Reverse(){
	if s.Size<=1{
		return
	}

	cursor := s.Head
	var prev *Node // 这里不要用 var prev = &Node{nil, nil}这种形式，因为这样prev!=nil是成立的
	for cursor != nil{
		current := cursor
		cursor = cursor.Next
		current.Next = prev
		prev = current
	}
	s.Head = prev
}

func (s *SingleChain) Plus(other *SingleChain) float64{
	var result float64 = 0
	for _,item := range []*SingleChain{s, other}	{
		count := 1
		var itemSum float64=0
		cursor := item.Head
		for cursor!=nil{
			itemSum += float64(cursor.Data)*math.Pow(float64(10), float64(item.Size-count))
			cursor = cursor.Next
			count++
		}
		result += itemSum
	}
	return result
}

func main(){
	s := &SingleChain{nil,0}
	s.Add(&Node{1, nil})
	s.Add(&Node{2, nil})
	s.Add(&Node{3, nil})
	fmt.Println(s)
	//fmt.Println("After reverse================")
	//s.Reverse()
	//fmt.Println(s)
	s1 := &SingleChain{nil,0}
	s1.Add(&Node{4, nil})
	s1.Add(&Node{5, nil})
	fmt.Println(s.Plus(s1))

	fmt.Println(s)
	fmt.Println(s1)
}

File: python/SingleChain.py
# encoding: utf-8
"""
    单链表
        解释: https://zh.wikipedia.org/wiki/%E5%8D%95%E5%90%91%E9%93%BE%E8%A1%A8
        借鉴: https://blog.csdn.net/Marksinoberg/article/details/69310033
"""
from copy import deepcopy

class Node(object):
    """
    存储节点
    """

    def __init__(self, data, _next=None):
        self.data = data
        self.next = _next


class SingleChain(object):
    """
    单链表
    """

    def __init__(self, head=None):
        self.head = head
        self.size = 0

    def append(self, value):
        """
        在后面添加一个元素
        :param value:
        :return:
        """
        new_node = Node(value)
        if self.head is None or self.size == 0:
            self.head = new_node
        else:
            cursor = self.head
            while cursor.next is not None:
                cursor = cursor.next
            cursor.next = new_node
        self.size += 1

    def __str__(self):
        cursor = self.head
        result = []
        while cursor is not None:
            result.append(cursor.data)
            cursor = cursor.next
        return "->".join([str(item) for item in result])

    def __getitem__(self, index):
        if index < 0 or index > self.size - 1:
            return None
        if self.head is None or self.size == 0:
            return None
        cursor = self.head
        count = 0
        while cursor is not None:
            if count == index:
                return cursor
            cursor = cursor.next
            count += 1

    def __setitem__(self, index, value):
        if index < 0 or index > self.size:
            return
        new_node = Node(value)
        self.size += 1
        if self.head is None or self.size == 0:
            self.head = new_node
            return
        if index == self.size:
            self[self.size - 1].next = new_node
            return
        if index == 0:
            old_head = self.head
            self.head = new_node
            new_node.next = old_head
        else:
            cursor = self[index]
            pre = self[index - 1]
            pre.next = new_node
            new_node.next = cursor

    def __iter__(self):
        self.cursor = self.head
        return self

    def next(self):
        """
        实现迭代
        :return:
        """
        temp = self.cursor
        if self.cursor is None:
            raise StopIteration()
        else:
            self.cursor = self.cursor.next
            return temp

    def __reversed__(self):
        """
        逆序
        :return:
        """
        if self.size <= 1:
            return
        prev = None
        cursor = self.head
        while cursor is not None:
            current = cursor
            cursor = cursor.next
            current.next = prev
            prev = current
        self.head = prev

    def __add__(self, other):
        """
        两个单链表相加, 1->2->3 + 4->5 = 168
        :param other:
        :return:
        """
        _self = deepcopy(self)
        _other = deepcopy(other)
        reversed(_self)
        reversed(_other)
        result = []
        for item in [self, other]:
            cursor = item.head
            count = 0
            add_result = 0
            while cursor is not None:
                add_result += cursor.data * (10 ** count)
                count += 1
                cursor = cursor.next
            result.append(add_result)
        return sum(result)


if __name__ == "__main__":
    s = SingleChain()
    s.append(1)
    s.append(2)
    s.append(3)
    print s.size
    print s
    print "取元素", "=" * 15
    print s[0].data
    print s[1].data
    print s[2].data
    print "设置元素", "=" * 15
    s[0] = -1
    s[1] = 11
    s[4] = 4
    print "反向", "=" * 15
    print s
    reversed(s)
    print s
    print "迭代", "=" * 15
    for item in s:
        print item.data

    print "相加", "=" * 15
    s1 = SingleChain()
    s1.append(1)
    s1.append(2)
    s1.append(3)
    s2 = SingleChain()
    s2.append(4)
    s2.append(5)
    print s2 + s1
